### PR TITLE
Theme Showcase: New component SearchThemes

### PR DESCRIPTION
**Pull request by taipeicoder**
#### Proposed Changes

This PR implements a new theme search experience as defined in 0EogtxvlmxMGcnEEaMVECv-fi-1259%3A37689.
Screen Shot 2022-11-09 at 4 36 17 PM

To accomplish this, we have:
1. Created a new component `<SearchThemes/>` and,
2. Updated the existing component `<KeyedSuggestions/>` to support new UX behaviors.

Note that the `<SearchThemes />` component is based on the existing component `<ThemesMagicSearchCard />` ([src](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/themes-magic-search-card/index.jsx)), and for the most part they should behave similar when it comes to text injection based on suggestion selection. 

The new search experience is as follows:

Whenever the search input field is clicked, the suggestion list will show up and display three main filters: "Feature", "Columns" and "Subject".
Screen Shot 2022-11-09 at 6 35 33 PM

Clicking on any of the above will refine the suggestion list to the list of terms from the selected filters. For instance, selecting "Subject" will inject the text `subject:` and update the suggestion list to terms under the filter "Subject", such as "Blog", "Business", etc.
Screen Shot 2022-11-09 at 6 35 51 PM

Clicking on a term will trigger search which updates the theme grid below the search bar. For instance, selecting "Blog" will update the text to `subject:blog` and close the suggestion list. 
Screen Shot 2022-11-09 at 6 36 19 PM

Unlike the old search experience, the suggestion list will no longer auto-suggest the first result. This is to allow users to do a plain-text search without being "forced" to select one of the suggestion. To understand more about this, try typing "blog" in the old search experience and do a plain-text search. Note that in the new search experience, keyboard navigation is still available.
Screen Shot 2022-11-09 at 6 36 44 PM

#### Testing Instructions

* Head to the theme showcase `/themes/${site_slug}`.
  * If using calypso.live, the flag `themes/showcase-i4/search-and-filter` is required.
* Ensure that the search resultset is the same as `wordpress.com/themes/${site_slug}`.
* Ensure that the search and suggestion behavior corresponds to the one described above.
* Ensure that the old search component works as it used to.
  * Disable the new component via flag, i.e.: `/themes/${site_slug}?flags=-themes/showcase-i4/search-and-filter`

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #69658

### CODE REVIEW

File: client/components/search-themes/index.tsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+	const onKeyDown = ( event: React.KeyboardEvent< HTMLInputElement > ) => {
+		suggestionsRef.current?.handleKeyEvent( event );
+	};
+
+	return (
+		<div
+			ref={ wrapperRef }
+			className={ classNames( 'search-themes', {
+				'has-keyed-suggestions': isSearchOpen,
```
Comment (arthur791004) [inline client/components/search-themes/index.tsx]: I guess we don't need this anymore 🤔

File: client/components/search-themes/style.scss [hunk 2]
```diff
@@ -0,0 +1,170 @@
+@import "@wordpress/base-styles/breakpoints";
+@import "@wordpress/base-styles/mixins";
+
+$search-border-radius: 4px;
+
+.search-themes-card {
+	display: flex;
+	align-items: center;
+	background: var(--color-surface);
+	transition: all 0.15s ease-in-out;
```
Comment (arthur791004) [inline client/components/search-themes/style.scss]: What's the purpose of this? I cannot see any usage 🤔

File: client/components/keyed-suggestions/index.jsx [hunk 3]
```diff
@@ -330,28 +373,59 @@ class KeyedSuggestions extends Component {
 		} );
 	};
 
+	createTopLevelTermsSuggestions = () => {
+		const rendered = [
```
Comment (arthur791004) [inline client/components/keyed-suggestions/index.jsx]: [nit] I think we could return the `JSX.Element` directly and we don't need to follow the `createSuggestions`.

File: client/components/keyed-suggestions/index.jsx [hunk 4]
```diff
@@ -7,6 +8,34 @@ import { cosineSimilarity } from 'calypso/lib/trigram';
 
 import './style.scss';
 
+const TAXONOMY_TRANSLATIONS = {
+	feature: i18n.translate( 'Feature', {
```
Comment (arthur791004) [inline client/components/keyed-suggestions/index.jsx:23]: The translation doesn't work in this way.

image

We have to define it inside the component or make a function to return the taxonomy as we cannot get locale when we import this file.

File: client/components/search-themes/use-outside-click-callback.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Hook that executes `callback` is the 'Escape' key is pressed
+ * or if the user clicks outside of `ref`.
+ *
+ * @param ref        Ref to an HTML element
+ * @param callback   Function to be executed
+ * @param useCapture useCapture flag passed to the add/removeEventListener
+ */
+export default function useOutsideClickCallback(
```
Comment (arthur791004) [inline client/components/search-themes/use-outside-click-callback.ts]: I think we can put it under the `lib/use-outside-click-callback` and share it with `jetpack/profile-dropdown` and her

File: client/components/search-themes/index.tsx (hunk 1 above)
Comment (taipeicoder) [reply client/components/search-themes/index.tsx]: You're right, I've removed it 🙂

File: client/components/search-themes/style.scss (hunk 2 above)
Comment (taipeicoder) [reply client/components/search-themes/style.scss]: Must have been carried over from the old component 😅 good catch!

File: client/components/keyed-suggestions/index.jsx (hunk 3 above)
Comment (taipeicoder) [reply client/components/keyed-suggestions/index.jsx]: Good idea, it looks cleaner!

File: client/components/keyed-suggestions/index.jsx (hunk 4 above)
Comment (taipeicoder) [reply client/components/keyed-suggestions/index.jsx:23]: You're right, thanks for pointing this out.

File: client/components/search-themes/use-outside-click-callback.ts (hunk 5 above)
Comment (taipeicoder) [reply client/components/search-themes/use-outside-click-callback.ts]: I'll create a follow-up issue to have the `jetpack/profile-dropdown` component to import from this file as well (as I'm not sure if it'd trigger code review from other teams, or other deployment processes).

Comment (arthur791004) [reply client/components/search-themes/use-outside-click-callback.ts]: I found another choice to use the `__experimentalUseFocusOutside` hook from Gutenberg

```js
import { __experimentalUseFocusOutside as useFocusOutside } from '@wordpress/compose';
```

Comment (taipeicoder) [reply client/components/search-themes/use-outside-click-callback.ts]: Changing to `useFocusOutside` seems to work nicely 👍

**Review by arthur791004 — APPROVED** [review]
Nice job 👍

**Comment by fushar** [comment]
Nice! I played around with the feature, and have some feedbacks / clarifications.

1. Looks like there should be a vertical space between the search box and the recommended / trending filter? Currently it looks a bit weird to me as the search box is rounded and the filter below it is not.

image

**Comment by fushar** [comment]
2. When I typed "blog" and hit Enter (to do a plaintext search), the suggestions are not closed and therefore obstruct the (first row of the) search results:

image

**Comment by fushar** [comment]
3. Is the text "all themes" constant? (Havent looked into the code in much details). For example I'm now on Trending themes tab (not all themes) but the placeholder still says "Search all themes", which looks buggy.

I see that in Saxon's Figma video, the placeholder text is "Search themes..."

image

**Comment by fushar** [comment]
4. Looks like the search box is intended to be sticky. However, it looks like it is composed of two rounded boxes (?), which could make the background to be visible in an awkward way. For example, see the below screenshot: the orange background from the Rainfall theme is "spilling over"

image

**Comment by fushar** [comment]
Other than the UI comments above, the experience feels really good and smooth! I will take a look into the code in more details tomorrow 👍

**Comment by taipeicoder** [comment]
> Looks like there should be a vertical space between the search box and the recommended / trending filter? Currently it looks a bit weird to me as the search box is rounded and the filter below it is not.

I plan to address it in the next PR, which is a revamp of the filter group below 🙂

>When I typed "blog" and hit Enter (to do a plaintext search), the suggestions are not closed and therefore obstruct the (first row of the) search results.

Nice catch! I've fixed it in the latest commit.

>Is the text "all themes" constant? (Havent looked into the code in much details). For example I'm now on Trending themes tab (not all themes) but the placeholder still says "Search all themes", which looks buggy.

You're right, I've changed it to "Search themes..." which is more generic.

>Looks like the search box is intended to be sticky. However, it looks like it is composed of two rounded boxes (?), which could make the background to be visible in an awkward way. For example, see the below screenshot: the orange background from the Rainfall theme is "spilling over"

Fixed, thank you for noticing this 🙏

**Review by fushar — APPROVED** [review]
Looks way better now! 😄

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7625182

Thank you @taipeicoder for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7625182).

**Pull request merged**